apps/minifront/src/components/swap/swap-form/price-history.tsx

```diff
@@ -0,0 +1,111 @@
+import { getMetadataFromBalancesResponseOptional } from '@penumbra-zone/getters/balances-response';
+import { AbridgedZQueryState } from '@penumbra-zone/zquery/src/types';
+import { Box } from '@repo/ui/components/ui/box';
+import { CandlestickPlot } from '@repo/ui/components/ui/candlestick-plot';
+import { useEffect, useMemo, useState } from 'react';
+import { getBlockDate } from '../../../fetchers/block-date';
+import { AllSlices } from '../../../state';
+import { useStatus } from '../../../state/status';
+import { combinedCandlestickDataSelector } from '../../../state/swap/helpers';
+import { useCandles, useRevalidateCandles } from '../../../state/swap/price-history';
+import { useStoreShallow } from '../../../utils/use-store-shallow';
+import { Button } from '@repo/ui/components/ui/button';
+
+const priceHistoryRequestSelector = (state: AllSlices) => ({
+  startMetadata: getMetadataFromBalancesResponseOptional(state.swap.assetIn),
+  endMetadata: state.swap.assetOut,
+  historyLimit: state.swap.priceHistory.historyLimit,
+  historyStart: state.swap.priceHistory.historyStart,
+});
+
+const latestKnownBlockHeightSelector = (
+  state: AbridgedZQueryState<{
+    fullSyncHeight?: bigint;
+    latestKnownBlockHeight?: bigint;
+  }>,
+) => state.data?.latestKnownBlockHeight;
+
+export const PriceHistory = () => {
+  const { startMetadata, endMetadata, historyLimit, historyStart } = useStoreShallow(
+    priceHistoryRequestSelector,
+  );
+  const latestKnownBlockHeight = useStatus({ select: latestKnownBlockHeightSelector });
+
+  const [lastCandlesUpdate, setLastCandlesUpdate] = useState<bigint>();
+  const [staleCandles, setStaleCandles] = useState<boolean>();
+  const [pause, setPause] = useState<bigint>();
+
+  const candles = useCandles(
+    { select: combinedCandlestickDataSelector },
+    startMetadata,
+    endMetadata,
+    historyLimit,
+    undefined,
+  );
+
+  const refetchCandles = useRevalidateCandles();
+
+  const candleRefetchArgs: Parameters<typeof refetchCandles> = useMemo(() => {
+    // args changes require refetch
+    setStaleCandles(true);
+    return [startMetadata, endMetadata, historyLimit, undefined];
+  }, [startMetadata, endMetadata, historyLimit]);
+
+  // auto-update and rate limit candlestick data
+  useEffect(() => {
+    if (!staleCandles && latestKnownBlockHeight && lastCandlesUpdate) {
+      setStaleCandles(latestKnownBlockHeight - lastCandlesUpdate > 10n);
+    }
+  }, [lastCandlesUpdate, latestKnownBlockHeight, pause, staleCandles]);
+
+  useEffect(() => {
+    if (staleCandles && !pause) {
+      setStaleCandles(false);
+      setLastCandlesUpdate(latestKnownBlockHeight);
+      refetchCandles(...candleRefetchArgs);
```

Comment:
Hm, I should probably have changed the `revalidate` function to return a promise, to allow you to do things like `setLastCandlesUpdate()` _after_ revalidation completes, but 🤷🏻‍♂️ 

More of a note to myself — nothing needs to change here